lib/active_record/base_without_table.rb

```diff
@@ -14,28 +14,40 @@ module ActiveRecord
   # This model can be used just like a regular model based on a table, except it will never be saved to the database.
   #
   class BaseWithoutTable < Base
-    self.abstract_class = true
-
-    class_attribute :columns
-    self.columns = []
+    attr_accessor :attributes
 
-    def create_or_update # :nodoc:
-      errors.empty?
-    end
+    self.abstract_class = true
 
     class << self
-      def column_defaults
-        columns.each_with_object(ActiveSupport::HashWithIndifferentAccess.new) do |current, res|
-          res[current.name] = current.default
-        end
+      def connection
+        Class.new(SimpleDelegator) do
+          def schema_cache
+            Class.new do
+              def columns(*args)
+                []
+              end
+
+              def columns_hash(*args)
+                {}
+              end
+            end.new
+          end
+        end.new(super)
       end
 
-      def columns_hash
-        all_columns = add_user_provided_columns(columns)
+      def table_exists?
+        false
+      end
 
-        @columns_hash = all_columns.each_with_object(ActiveSupport::HashWithIndifferentAccess.new) do |current, res|
-          res[current.name] = current
+      def inherited(subclass)
```

Comment:
As discussed, we can raise if a programmer attempts to inherit a class that inherits from base without table.